### DISCUSSION TITLE
Can MOOSE output STL files of the system under consideration?

### DISCUSSION THREAD

**Discussion by kurtenkera** [idaholab/moose#27023] [Q&A General]
I'm developing a project where I'd like MOOSE to output an STL file of the system under consideration? I will start by using a .msh file created with Gmsh, but then I intend to perform topology optimisation on the system, and I want to output an STL file of the topology optimised component. Is this possible?

**Comment by GiudGiud** [comment]
Hello

We dont have this capability.
I think reading a STL then meshing it is more within our interests currently than outputting STLs.
So if you want to do this, you ll need to implement it yourself.

Guillaume

**Reply by kurtenkera** [reply]
Thanks for your swift response. So would I at least be able to get the .msh file of the outputted component and then you’re saying I’d have to produce some code that converts the outputted .msh file into an STL file? Where can I find a list of mesh file types (e.g., .msh) outputted by MOOSE so I can potentially explore other options if converting from .msh to STL is not possible?

**Reply by GiudGiud** [reply]
https://mooseframework.inl.gov/source/meshgenerators/FileMeshGenerator.html has a list

**Reply by GiudGiud** [reply]
I added it for convenience but otherwise there s another list under the Outputs syntax

**Reply by kurtenkera** [reply]
So just to be clear, I should be able to write my final topology optimized component to a .msh ASCII file (as suggested below)? I'm unsure as below it says that 'these formats... could be easily **added** to MOOSE' - does this mean I'd have to somehow do some extra steps to **add** the capability of writing a .msh ASCII file in MOOSE?

![image](https://github.com/idaholab/moose/assets/127472392/d0b69017-eae3-492f-a1de-948f112e7e3a)

**Reply by GiudGiud** [reply]
Hello

Yes libmesh can write with these formats but moose cannot for now. So you need to create a new MOOSE output object that uses the libmesh APIs to trigger an output of the moose mesh to a gmsh .msh. 

This should be fairly straightforward to do but it s C++ work so you should have followed one of the development tutorials 

Guillaume

**Reply by kurtenkera** [reply]
@GiudGiud you mentioned that reading in an STL and meshing it is something being investigated by MOOSE. I'm currently trying to convert a binary STL file to a MSH ASCII file using the 'gmsh'/'meshio' libraries but not having any luck - do you know a way around this problem?

**Reply by GiudGiud** [reply]
Hello

STL is not usually convertible to a mesh file without specialized meshing software. i think gmsh should be able to do it. otherwise, you can use Cubit. If you work for a university there is a student version (of coreform cubit, a private company's product), if you work for a US government contractor, there is a release version (of regular cubit) by Sandia National Lab. 

Guillaume

**Reply by kurtenkera** [reply]
Hi @GiudGiud, 

You said the following above:

> Yes libmesh can write with these formats but moose cannot for now. So you need to create a new MOOSE output object that uses the libmesh APIs to trigger an output of the moose mesh to a gmsh .msh.

Is there an easy to follow example where someone has done this (i.e., created a MOOSE app that uses the libmesh APIs to trigger an output of the MOOSE mesh to a gmsh .msh file)? Or would my best bet be to follow the 11hr tutorial closely (particularly the sections on adding functionality to MOOSE via c++)?

**Reply by GiudGiud** [reply]
Hello

There isnt a tutorial on creating a new Outputs object.
However, Exodus, Tecplot, etc most of the outputs are exactly that, a MOOSE object using libmesh APIs.

The code in libmesh is there
https://github.com/libMesh/libmesh/blob/devel/src/mesh/gmsh_io.C
we just need to call it from MOOSE

Guillaume

**Reply by kurtenkera** [reply]
@GiudGiud where can I find the header files that are used to define Exodus/Tecplot MOOSE objects? I'm hoping I can copy the C++ code for these header files and make a few (hopefully simple) changes so that I can create a header file that allows my MOOSE-App to output .msh files?

**Reply by GiudGiud** [reply]
The source is here
https://github.com/idaholab/moose/tree/next/framework/src/outputs
the headers are here
https://github.com/idaholab/moose/tree/next/framework/include/outputs

> 'm hoping I can copy the C++ code for these header files and make a few (hopefully simple) changes so that I can create a header file that allows my MOOSE-App to output .msh files?

some of these output classes will be closer than others. exodus might be on the more complex side, maybe tecplot or vtk are much closer to what you need

**Reply by kurtenkera** [reply]
Much appreciated! So if I create my own .c and .h files that allow my MOOSE-app to output a .msh file, I would have to append these files to the local directories `~/projects/moose/tree/next/framework/src/outputs` and `~/projects/moose/tree/next/framework/include/outputs` respectively? And if I get the code working, I can potentially make a pull request to the `moose` source code if this implementation is successful?

**Reply by GiudGiud** [reply]
Hello

.C instead of .c 
But otherwise yes for both 

Guillaume
